File: csrc/dsa/torch_ops_extension/custom_ops/converter/npu_sparse_flash_attention.py
# This program is free software, you can redistribute it and/or modify it.
# Copyright (c) 2025 Huawei Technologies Co., Ltd.
# This file is a part of the CANN Open Software.
# Licensed under CANN Open Software License Agreement Version 2.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.

from typing import (
    Any,
    Callable,
    ContextManager,
    Iterator,
    List,
    Literal,
    NamedTuple,
    Optional,
    Sequence,
    Tuple,
    TypeVar,
    Union,
    overload,
)

import torch
import torchair
from torch import Generator, SymInt, contiguous_format, inf, strided
from torch.types import (
    Device,
    Number,
    _bool,
    _complex,
    _device,
    _dtype,
    _float,
    _int,
    _layout,
    _qscheme,
    _size,
)
from torchair._ge_concrete_graph import ge_apis as ge
from torchair._ge_concrete_graph.fx2ge_converter import (
    declare_supported,
    register_fx_node_ge_converter,
)
from torchair._ge_concrete_graph.supported_declaration import (
    BOOL,
    F16,
    F32,
    F64,
    I8,
    I16,
    I32,
    I64,
    U8,
    Support,
    _TypedTensor,
)
from torchair.ge import attr
from torchair.ge._ge_graph import DataType, Tensor, TensorSpec


# 为自定义算子注册converter，用于torch.compile 场景成图
# 注意： meta_outputs形参名为固定写法，若写错会影响ge节点的输出dtype与shape推导
@register_fx_node_ge_converter(torch.ops.custom.npu_sparse_flash_attention.default)
def convert_npu_sparse_flash_attention(
    query: Tensor,
    key: Tensor,
    value: Tensor,
    sparse_indices: Tensor,
    scale_value: float,
    sparse_block_size: int,
    *,
    block_table: Optional[Tensor] = None,
    actual_seq_lengths_query: Optional[Tensor] = None,
    actual_seq_lengths_kv: Optional[Tensor] = None,
    query_rope: Optional[Tensor] = None,
    key_rope: Optional[Tensor] = None,
    layout_query: str = "BSND",
    layout_kv: str = "BSND",
    sparse_mode: int = 3,
    meta_outputs: TensorSpec = None,
):
    return torchair.ge.custom_op(
        "SparseFlashAttention",
        inputs={
            "query": query,
            "key": key,
            "value": value,
            "sparse_indices": sparse_indices,
            "block_table": block_table,
            "actual_seq_lengths_query": actual_seq_lengths_query,
            "actual_seq_lengths_kv": actual_seq_lengths_kv,
            "query_rope": query_rope,
            "key_rope": key_rope,
        },
        attrs={
            "scale_value": attr.Float(scale_value),
            "sparse_block_size": attr.Int(sparse_block_size),
            "layout_query": attr.Str(layout_query),
            "layout_kv": attr.Str(layout_kv),
            "sparse_mode": attr.Int(sparse_mode),
        },
        outputs=["attention_out"],
    )

File: csrc/dsa/torch_ops_extension/custom_ops/csrc/npu_sparse_flash_attention.cpp
/**
 * This program is free software, you can redistribute it and/or modify it.
 * Copyright (c) 2025 Huawei Technologies Co., Ltd.
 * This file is a part of the CANN Open Software.
 * Licensed under CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING
 * BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE. See LICENSE in the root of
 * the software repository for the full text of the License.
 */

#include <iostream>
#include <torch/library.h>
#include "ops_common.h"

namespace custom {
using namespace at_npu::native;

// 工具函数，推导输出shape
at::Tensor construct_sparse_infer_output_tensor(const at::Tensor &query, const at::Tensor &value, std::string layout)
{
    at::SmallVector<int64_t, 8> output_size;
    if (layout == "TND") {
        output_size = {query.size(DIM_0), query.size(DIM_1), query.size(DIM_2)};
    } else {
        output_size = {query.size(DIM_0), query.size(DIM_1), query.size(DIM_2), query.size(DIM_3)};
    }
    at::Tensor output = at::empty(output_size, query.options().dtype(query.dtype()));

    return output;
}

// step2, 为NPU设备实现前向接口
at::Tensor npu_sparse_flash_attention_npu(const at::Tensor &query, const at::Tensor &key, const at::Tensor &value,
                                          const at::Tensor &sparse_indices, double scale_value,
                                          int64_t sparse_block_size, const c10::optional<at::Tensor> &block_table,
                                          const c10::optional<at::Tensor> &actual_seq_lengths_query,
                                          const c10::optional<at::Tensor> &actual_seq_lengths_kv,
                                          const c10::optional<at::Tensor> &query_rope,
                                          const c10::optional<at::Tensor> &key_rope, c10::string_view layout_query,
                                          c10::string_view layout_kv, int64_t sparse_mode)
{
    std::string layout_query_str = std::string(layout_query);
    std::string layout_kv_str = std::string(layout_kv);

    // construct the output tensor
    at::Tensor output = construct_sparse_infer_output_tensor(query, value, layout_query_str);
    // convert str
    char *layout_query_ptr = const_cast<char *>(layout_query_str.c_str());
    char *layout_kv_ptr = const_cast<char *>(layout_kv_str.c_str());

    EXEC_NPU_CMD_V1(aclnnSparseFlashAttention, query, key, value, sparse_indices, block_table, actual_seq_lengths_query,
                    actual_seq_lengths_kv, query_rope, key_rope, scale_value, sparse_block_size, layout_query_ptr,
                    layout_kv_ptr, sparse_mode, output);
    return output;
}

// step3, 为META设备实现前向接口
at::Tensor npu_sparse_flash_attention_meta(const at::Tensor &query, const at::Tensor &key, const at::Tensor &value,
                                           const at::Tensor &sparse_indices, double scale_value,
                                           int64_t sparse_block_size, const c10::optional<at::Tensor> &block_table,
                                           const c10::optional<at::Tensor> &actual_seq_lengths_query,
                                           const c10::optional<at::Tensor> &actual_seq_lengths_kv,
                                           const c10::optional<at::Tensor> &query_rope,
                                           const c10::optional<at::Tensor> &key_rope, c10::string_view layout_query,
                                           c10::string_view layout_kv, int64_t sparse_mode)
{
    std::string layout_query_str = std::string(layout_query);
    at::Tensor output = construct_sparse_infer_output_tensor(query, value, layout_query_str);

    return output;
}
}  // namespace custom

// step4, 为NPU设备注册前向实现
TORCH_LIBRARY_IMPL(custom, PrivateUse1, m)
{
    m.impl("npu_sparse_flash_attention", &custom::npu_sparse_flash_attention_npu);
}

// step5, 为META设备注册前向实现
TORCH_LIBRARY_IMPL(custom, Meta, m)
{
    m.impl("npu_sparse_flash_attention", &custom::npu_sparse_flash_attention_meta);
}
